x-pack/plugins/ai_infra/product_doc_base/server/services/package_installer/steps/populate_index.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { BulkRequest } from '@elastic/elasticsearch/lib/api/types';
+import type { Logger } from '@kbn/logging';
+import type { ElasticsearchClient } from '@kbn/core/server';
+import { isArtifactContentFilePath } from '@kbn/product-doc-common';
+import { internalElserInferenceId } from '../../../../common/consts';
+import type { ZipArchive } from '../utils/zip_archive';
+
+export const populateIndex = async ({
+  esClient,
+  indexName,
+  archive,
+  log,
+}: {
+  esClient: ElasticsearchClient;
+  indexName: string;
+  archive: ZipArchive;
+  log: Logger;
+}) => {
+  log.debug(`Starting populating index ${indexName}`);
+
+  const contentEntries = archive.getEntryPaths().filter(isArtifactContentFilePath);
+
+  for (let i = 0; i < contentEntries.length; i++) {
+    const entryPath = contentEntries[i];
+    log.debug(`Indexing content for entry ${entryPath}`);
+    const contentBuffer = await archive.getEntryContent(entryPath);
+    await indexContentFile({ indexName, esClient, contentBuffer });
+  }
+
+  log.debug(`Done populating index ${indexName}`);
+};
+
+const indexContentFile = async ({
+  indexName,
+  contentBuffer,
+  esClient,
+}: {
+  indexName: string;
+  contentBuffer: Buffer;
+  esClient: ElasticsearchClient;
+}) => {
+  const fileContent = contentBuffer.toString('utf-8');
+  const lines = fileContent.split('\n');
+
+  const documents = lines
+    .map((line) => line.trim())
+    .filter((line) => line.length > 0)
+    .map((line) => {
+      return JSON.parse(line);
+    })
+    .map((doc) => rewriteInferenceId(doc, internalElserInferenceId));
+
+  const operations = documents.reduce((ops, document) => {
+    ops!.push(...[{ index: { _index: indexName } }, document]);
+    return ops;
+  }, [] as BulkRequest['operations']);
+
+  const response = await esClient.bulk({
+    refresh: false,
+    operations,
+  });
+
+  if (response.errors) {
+    const error = response.items.find((item) => item.index!.error)!.index!.error!;
```

Comment:
Will `item.index` always be defined? I'd hate to hit an error while generating an error

Reply:
Yeah it will, we're only performing `index` operations, so the response items will all have an `index` property.

But I'm still surprised I used `!` instead of `?` here tbh, I usually still play it cautious, I can adapt.